pstoreds/addr_book.go

```diff
@@ -79,141 +98,206 @@ func (mgr *dsAddrBook) AddAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Durati
+	}
+	mgr.setAddrs(p, addrs, ttl, true)
 }
 
-func (mgr *dsAddrBook) setAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration, add bool) {
-	for i := 0; i < dsWriteRetries; i++ {
-		// keys to add to the TTL manager
-		var keys []ds.Key
-		batch, err := mgr.ds.Batch()
-		if err != nil {
-			log.Error(err)
-			return
+func (mgr *dsAddrBook) deleteAddrs(p peer.ID, addrs []ma.Multiaddr) error {
+	// Keys and cleaned up addresses.
+	keys, addrs, err := keysAndAddrs(p, addrs)
+	if err != nil {
+		return err
+	}
+
+	mgr.cache.Remove(p.Pretty())
+	// Attempt transactional KV deletion.
+	for i := 0; i < mgr.writeRetries; i++ {
+		if err = mgr.dbDelete(keys); err == nil {
+			break
 		}
+		log.Errorf("failed to delete addresses for peer %s: %s\n", p.Pretty(), err)
+	}
 
-		for _, addr := range addrs {
-			if addr == nil {
-				continue
-			}
+	if err != nil {
+		log.Errorf("failed to avoid write conflict for peer %s after %d retries: %v\n", p.Pretty(), mgr.writeRetries, err)
+		return err
+	}
 
-			key, err := peerAddressKey(&p, &addr)
-			if err != nil {
-				log.Error(err)
-				continue
-			}
-			keys = append(keys, key)
+	mgr.ttlManager.deleteTTLs(keys)
+	return nil
+}
 
-			if ttl <= 0 {
-				if err := batch.Delete(key); err != nil {
-					log.Error(err)
-				} else {
-					mgr.cache.Remove(key)
-				}
-				continue
-			}
+func (mgr *dsAddrBook) setAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration, ttlReset bool) error {
+	// Keys and cleaned up addresses.
+	keys, addrs, err := keysAndAddrs(p, addrs)
+	if err != nil {
+		return err
+	}
 
-			has := mgr.cache.Contains(key)
-			if !has {
-				has, err = mgr.ds.Has(key)
-			}
-			if err != nil || !has {
-				mgr.subsManager.BroadcastAddr(p, addr)
-			}
+	mgr.cache.Remove(p.Pretty())
+	// Attempt transactional KV insertion.
+	var existed []bool
+	for i := 0; i < mgr.writeRetries; i++ {
+		if existed, err = mgr.dbInsert(keys, addrs); err == nil {
+			break
+		}
+		log.Errorf("failed to write addresses for peer %s: %s\n", p.Pretty(), err)
+	}
 
-			// Allows us to support AddAddr and SetAddr in one function
-			if !has {
-				if err := batch.Put(key, addr.Bytes()); err != nil {
-					log.Error(err)
-				} else {
-					mgr.cache.Add(key, addr.Bytes())
-				}
-			}
+	if err != nil {
+		log.Errorf("failed to avoid write conflict for peer %s after %d retries: %v\n", p.Pretty(), mgr.writeRetries, err)
+		return err
+	}
+
+	// Successful. Update cache and broadcast event.
+	for i, _ := range keys {
+		if !existed[i] {
+			mgr.subsManager.BroadcastAddr(p, addrs[i])
+		}
+	}
+
+	// Force update TTLs only if TTL reset was requested; otherwise
+	// insert the appropriate TTL entries if they don't already exist.
+	if ttlReset {
+		mgr.ttlManager.setTTLs(keys, ttl)
+	} else {
+		mgr.ttlManager.insertTTLs(keys, ttl)
+	}
+
+	return nil
+}
+
+// dbInsert performs a transactional insert of the provided keys and values.
+func (mgr *dsAddrBook) dbInsert(keys []ds.Key, addrs []ma.Multiaddr) ([]bool, error) {
```

Comment:
These private methods (dbInsert, dbClear, dbDelete) perform a single attempt, and the calling method performs as many retries as configured. Given that entries are content-addressed `(peerId/hash(value)) => value`, writes _really_ cannot conflict, so I'd agree that whole need for transactionality is somewhat disputed.